### PR TITLE
refactor(virtual_file) make write_all_at take owned buffers

### PR DESCRIPTION
**Pull request by problame**
context: https://github.com/neondatabase/neon/issues/6663

Building atop #6664, this PR switches `write_all_at` to take owned buffers.

The main challenge here is the `EphemeralFile::mutable_tail`, for which I'm picking the ugly solution of an `Option` that is `None` while the IO is in flight.

After this, we will be able to switch `write_at` to take owned buffers and call tokio-epoll-uring's `write` function with that owned buffer. That'll be done in #6378.

### CODE REVIEW

File: pageserver/src/virtual_file.rs [hunk 1]
```diff
@@ -582,24 +582,37 @@ impl VirtualFile {
     }
 
     // Copied from https://doc.rust-lang.org/1.72.0/src/std/os/unix/fs.rs.html#219-235
-    pub async fn write_all_at(&self, mut buf: &[u8], mut offset: u64) -> Result<(), Error> {
+    pub async fn write_all_at<B: BoundedBuf>(
+        &self,
+        buf: B,
+        mut offset: u64,
+    ) -> (B::Buf, Result<(), Error>) {
```
Comment (VladLazar) [inline pageserver/src/virtual_file.rs:589]: nit: could return `Result<B::Buf, Error>` to force callers to inspect the error.

**Review by VladLazar — APPROVED** [review]
Looks correct to me- just nits/suggestions

File: pageserver/src/virtual_file.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/virtual_file.rs:589]: The `Result` is `#[must_use]`, so, rustc should still warn if you don't handle the error.
I don't want to change the convention I stuck to throughout this whole patch series: you put in resources and you always get them back, along with a result.

File: pageserver/src/tenant/ephemeral_file.rs [hunk 2]
```diff
@@ -135,21 +145,27 @@ impl EphemeralFile {
...
-                                &self.ephemeral_file.mutable_tail,
-                                self.blknum as u64 * PAGE_SZ as u64,
-                            )
-                            .await
-                        {
+                            .write_all_at(mutable_tail, self.blknum as u64 * PAGE_SZ as u64)
+                            .await;
+                        // TODO: If we panic before we can put the mutable_tail back, subsequent calls will fail.
+                        // I.e., the IO isn't retryable if we panic.
+                        self.ephemeral_file.mutable_tail = Some(mutable_tail);
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/ephemeral_file.rs:167]: The `defer!` can't magically escape ownership rules.
We moved the mutable_tail into the write_all_at future.
If it panics, the `scopeguard::defer!` can't get it out.

File: pageserver/src/virtual_file.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/virtual_file.rs:589]: Ohhhh, actually `#[must_use]` doesn't work on tuples 🧠  https://github.com/rust-lang/rust/issues/39524

Comment (problame) [reply pageserver/src/virtual_file.rs:589]: I'll create a follow-up issue that I'll tend to after merging the patch series.

Comment (problame) [reply pageserver/src/virtual_file.rs:589]: https://github.com/neondatabase/neon/issues/6754

File: pageserver/src/tenant/ephemeral_file.rs [hunk 3]
```diff
@@ -26,7 +27,10 @@ pub struct EphemeralFile {
     /// An ephemeral file is append-only.
     /// We keep the last page, which can still be modified, in [`Self::mutable_tail`].
     /// The other pages, which can no longer be modified, are accessed through the page cache.
-    mutable_tail: [u8; PAGE_SZ],
+    ///
+    /// None <=> IO is ongoing.
+    /// Size is al
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/ephemeral_file.rs]: https://github.com/neondatabase/neon/pull/6673/commits/6a55ac7af50dfe380497d098a35768dc573cd7be

**Pull request merged**